lib/imagefs.coffee

```diff
@@ -23,16 +23,71 @@ _ = require('lodash')
 filedisk = require('file-disk')
 fs = Promise.promisifyAll(require('fs'))
 replaceStream = require('replacestream')
+ext2fs = require('ext2fs')
+
 driver = require('./driver')
+utils = require('./utils')
 
 checkImageType = (image) ->
 	if not (_.isString(image) or image instanceof filedisk.Disk)
 		throw new Error('image must be a String (file path) or a filedisk.Disk instance')
 
+composeDisposers = (outerDisposer, createInnerDisposer) ->
+	Promise.resolve(outerDisposer)
+	.then (outerDisposer) ->
+		outerDisposer._promise
+		.then (outerResult) ->
+			Promise.resolve(createInnerDisposer(outerResult))
+			.then (innerDisposer) ->
+				innerDisposer._promise
+				.then (innerResult) ->
+					Promise.resolve(innerResult)
+					.disposer (innerResult) ->
+						Promise.resolve(innerDisposer._data(innerResult))
+						.then ->
+							outerDisposer._data(outerResult)
+			.catch (err) ->
+				outerDisposer._data(outerResult)
+				throw err
+
+###*
+# @summary Get a bluebird.disposer of a node fs like interface for a partition
+# @function
+# @public
+#
+# @param {String|filedisk.Disk} disk - path to the image or filedisk.Disk instance
+# @param {Object} partition - partition definition
+#
+# @returns {bluebird.disposer<fs>} node fs like interface
+#
+# @example
+#
+# Promise.using imagefs.interact('/foo/bar.img', primary: 4, logical: 1), (fs) ->
+#   fs.readFileAsync('/bar/qux')
+#   .then (contents) ->
+#     console.log(contents)
+###
+exports.interact = (disk, partition) ->
+	checkImageType(disk)
+	if _.isString(disk)
+		composeDisposers(
+			filedisk.openFile(disk, 'r+')
+			(fd) ->
+				disk = new filedisk.FileDisk(fd, true)
+				driver.interact(disk, partition)
+		)
+	else if disk instanceof filedisk.Disk
+		driver.interact(disk, partition)
+
 read = (disk, partition, path) ->
-	driver.interact(disk, partition)
-	.then (fat) ->
-		fat.createReadStream(path)
+	composeDisposers(
+		driver.interact(disk, partition)
+		(fs_) ->
+			Promise.resolve(fs_.createReadStream(path, autoClose: false))
+			.disposer (stream) ->
+				if stream.fd != undefined
+					fs_.closeAsync(stream.fd)
```

Comment:
In that case you'll want to use `stream.close()`, which does accept a callback. You really don't want to close the `fd` from outside the stream – that'll interfere with the stream's internal state – setting `stream.destroyed = true` has no effect.